strawberryfields/engine.py

```diff
@@ -385,4 +420,154 @@ def run(self, program, *, compile_options=None, run_options=None):
+
+    def _init_backend(self, *args):
+        """
+        TODO: This does not do anything right now.
+        """
+        # Do nothing for now...
+        pass
+
+    def reset(self):
+        """
+        Reset must be called in order to submit a new job. This clears the job queue as well as
+        any ran Programs.
+        """
+        super().reset(backend_options={})
+        self.jobs.clear()
+
+    def _get_blackbird(self, name, shots, program):
+        """
+        Returns a Blackbird object to be sent later to the server when creating a job.
+        Assumes the current backend as the target.
+
+        Args:
+            name (str): the name of the job to be created (e.g. StateTeleportation)
+            shots (int): the number of shots
+            program (Program): program to be converted to Blackbird code
+
+        Returns:
+            blackbird.BlackbirdProgram
+        """
+        bb = to_blackbird(program, version="1.0")
+        bb._name = name
+
+        # TODO: This is potentially not needed here
+        bb._target["name"] = self.backend_name
+
+        bb._target["options"] = {"shots": shots}
+        return bb
+
+    def _queue_job(self, job_content):
+        """
+        Create a Job instance based on job_content, and send the job to the API. Append to list
+        of jobs.
+
+        Args:
+            job_content (str): the Blackbird code to execute
+
+        Returns:
+            (strawberryfields.api_client.Job): a Job instance referencing the queued job
+        """
+        job = Job(client=self.client)
+        job.manager.create(circuit=job_content)
+        self.jobs.append(job)
+        return job
+
+    def _run_program(self, program, **kwargs):
+        """
+        Given a compiled program, gets the blackbird circuit code and creates (or resumes) a job
+        via the API. If the job is completed, returns the job result.
+
+        A queued job can be interrupted by a ``KeyboardInterrupt`` event, at which point if the
+        job ID was retrieved from the server, the job will be accessible via
+        :meth:`~.Starship.jobs`.
+
+        Args:
+            program (strawberryfields.program.Program): program to be executed remotely
+
+        Returns:
+            (list): a list representing the result samples
+
+        Raises:
+            Exception: In case a job could not be submitted or completed.
+            TypeError: In case a job is already queued and a user is trying to submit a new job.
+        """
+        if self.jobs:
+            raise OneJobAtATimeError(
+                "A job is already queued. Please reset the engine and try again."
+            )
+
+        job_content = self._get_blackbird(program=program, **kwargs).serialize()
+        job = self._queue_job(job_content)
+
+        try:
+            while not job.is_failed and not job.is_complete:
+                job.reload()
+                sleep(self.polling_delay_seconds)
+        except KeyboardInterrupt:
```

Comment:
Great question -- I think that some refactoring is needed here. A job without an ID probably should never be added to `self.jobs`. I believe with the current setup this will occur (as an exception will be raised if the server does not respond with a `201`). I do agree that this can be coded better though.